concent_api/core/models.py

```diff
@@ -73,3 +76,167 @@ class MessageAuth(Model):
     message              = OneToOneField(StoredMessage, related_name = 'auth')
     provider_public_key  = Base64Field(max_length=64)
     requestor_public_key = Base64Field(max_length=64)
+
+
+class ClientManager(Manager):
+
+    def get_or_create_full_clean(self, public_key):
+        """
+        Returns Model instance.
+        Does the same as get_or_create method, but also performs full_clean() on newly created instance.
+        """
+        try:
+            instance = self.get(public_key = base64.b64encode(public_key))
+        except self.model.DoesNotExist:
+            instance = self.model(
+                public_key_bytes = public_key
+            )
+            instance.full_clean()
+            instance.save()
+        return instance
+
+
+class Client(models.Model):
+    """
+    Represents Concent client, identified by public key.
+    """
+
+    objects = ClientManager()
+
+    public_key = Base64Field(max_length = 64, unique = True)
+
+
+class Subtask(models.Model):
+    """
+    Represents subtasks states.
+    """
+
+    class SubtaskState(ChoiceEnum):
+        FORCING_REPORT          = 'forcing_report'
+        REPORTED                = 'reported'
+        FORCING_RESULT_TRANSFER = 'forcing_result_transfer'
+        RESULT_UPLOADED         = 'result_uploaded'
+        FORCING_ACCEPTANCE      = 'forcing_acceptance'
+        REJECTED                = 'rejected'
+        ADDITIONAL_VERIFICATION = 'additional_verification'
+        ACCEPTED                = 'accepted'
+        FAILED                  = 'failed'
+
+    # Defines Subtask model active states
+    SUBTASK_ACTIVE_STATES = [
+        SubtaskState.FORCING_REPORT.name,  # pylint: disable=no-member
+        SubtaskState.FORCING_RESULT_TRANSFER.name,  # pylint: disable=no-member
+        SubtaskState.FORCING_ACCEPTANCE.name,  # pylint: disable=no-member
+        SubtaskState.ADDITIONAL_VERIFICATION.name,  # pylint: disable=no-member
+    ]
+
+    # Defines Subtask model passive states
+    SUBTASK_PASSIVE_STATES = [
+        SubtaskState.REPORTED.name,  # pylint: disable=no-member
+        SubtaskState.RESULT_UPLOADED.name,  # pylint: disable=no-member
+        SubtaskState.REJECTED.name,  # pylint: disable=no-member
+        SubtaskState.ACCEPTED.name,  # pylint: disable=no-member
+        SubtaskState.FAILED.name,  # pylint: disable=no-member
+    ]
+
+    # Defines possible state transitions, where keys are transitions to, and values are lists of transitions from.
+    # None means transition from Unknown state.
+    POSSIBLE_STATE_TRANSITIONS = {
```

Comment:
See my comment above, it works, there are dozens of tests which go through that line.

Reply:
My bad - I initialy thought that state and _cuurent_state are of type `SubtaskState`, whereas they are of type `string` ;)

Reply:
Maybe the variables could be renamed to:
`state_name` and `_current_state_name`? Or better - use mypy annotation to indicate the type of variables?

Reply:
I agree with `_current_state_name` but `state` is actually a choice field so I would like to keep it this way. Plus, it is actually used in a lot of places now so changing this at this point would cause extra work, if we really want to do this lets create a separate issue out of it. But as I said, I would like to keep it this way.